### ISSUE TITLE
pgx v5 alpha 4 available for testing

### ISSUE THREAD

**Issue by jackc** [jackc/pgx#1169]
I'm pleased to announce the first alpha release of pgx v5 is available.

The code is in branch [v5-dev](https://github.com/jackc/pgx/tree/v5-dev). Check out the [changelog](https://github.com/jackc/pgx/blob/v5-dev/CHANGELOG.md) to see what's new.

There is a bit of missing functionality here and there, but it is complete enough I have been using it in production for a month or two. I don't recommend that to anyone else of course 😉. But I would appreciate it if people would start testing it.

Try it with:

```
go get github.com/jackc/pgx/v5@v5.0.0-alpha.1
```

**Comment by cristaloleg** [comment]
Any tldr what was changed?

**Comment by jackc** [comment]
Restructure of the type encoding and decoding system is the biggest change.

Beyond that there are a number of changes, but for the most part they involve various advanced usages. I hope that most casual users won't notice anything different.

**Comment by henvic** [comment]
Hey, just wanted to say that I'm happy to see pgtype, pgconn, and pgproto3 included in the Go module and the reduced dependency tree!

Navigating the codebase having to go to fewer repositories seems much easier now.

**Comment by tsingson** [comment]
v5 work fine in my v4 project, that only need change is logger adapter to uber-go/zap.

**Comment by jackc** [comment]
Good to hear!

Just pushed `v5.0.0-alpha.2`. Mostly just bug fixes and improved testing.

**Comment by jackc** [comment]
Just added `v5.0.0-alpha.3`.

* Now uses generic types for array, range, and multirange.
* Added named query argument support via `NamedArgs` and `QueryRewriter`.
* Merged in recent changes from v4
* Misc. small changes and bug fixes

**Comment by cristaloleg** [comment]
> Now uses generic types for array, range, and multirange.

So, Go mod will be 1.18 for v5?

**Comment by jackc** [comment]
> So, Go mod will be 1.18 for v5?

Yes.

**Comment by ecerichter** [comment]
Maybe I've found a corner case for named parameters; shall I open a new issue or reopen #1186?

**Comment by mcdoker18** [comment]
@jackc 

Right now all pg types have knowledge about their nullability.
If you are going to use generics and 1.18 for v5, did you think about the separate Nullable type?

```go
// or T must implement some interface.
type Nullable[T any] struct {
	T
	Valid  bool
}
```

At first glance, this can improve both the user experience and simplify the implementation.

**Comment by jackc** [comment]
Generics have worked well so far for array support and I've considered generics for nullable types. However, I've run into a few issues.

This would presumably lead to `NullT` and `T` types. This would essentially renamed all existing pgtype types `T` to `NullT`. Changing this could significantly add to the effort to upgrade applications to `v5`.

To gain code savings `Nullable[T]` would have to implement the null handling code and delegate to `T`. That means the value field would always be named the same. e.g. `NullText` would have field `Value string` instead of `String string`. This would also break existing code (and `Value()` is used for database/sql compatibility so we'd actually need a different name. Also, some `T` are native Go types. e.g. `string`. This change would require a `pgtype.Text` that implements the actual non-null logic so it could be used as `T` in `Nullable[T]`. But then that value field would be a `pgtype.Text` instead of `string`. That would make it harder to use.

If there is an approach that resolves these problems I would be interested in pursuing it, but I did not find it when I investigated.

**Comment by jackc** [comment]
Just tagged `v5.0.0-alpha.4`. Here's what's changed since alpha 3.

- `Trace()` method adds low level message tracing similar to the `PQtrace` function in `libpq`.
- `CheckConn()` checks a connection's liveness by doing a non-blocking read. This can be used to detect database restarts or network interruptions without executing a query or a ping.
- Various minor fixes and tweaks
- pgconn now uses non-blocking IO

The last item deserves special attention. It was necessary to efficiently implement `CheckConn()` and will be necessary for a future features such as tmaking context deadline exceeded not be fatal to a connection (i.e. adding a recover connection that can be called when it is in a bad state).

Non-blocking IO shouldn't make a visible difference. But due to Go not exposing non-blocking IO except through raw syscalls and TLS connections considering any write error to be fatal, the implementation for this is quite complex and subtle. It was much harder than it should have been. So there definitely could be bugs lurking there.

**Comment by jackc** [comment]
Just released `v5.0.0-alpha.5` and it's a pretty big release. Here are the major changes since alpha 4:

* Add `CollectRows` and `RowTo*` generic functions. These can **dramatically** reduce the boilerplate when reading multiple rows.
* pgx batch supports callback functions for more ergonomic usage. It's much easier to use batches when the result reading code is next to the query.
* pgconn now supports pipeline mode. This requires careful usage but can dramatically improve performance and / or reduce the number of connections required. It also enabled the following `SendBatch` performance improvement.
* `SendBatch` now uses only one network round trip to prepare / describe statements instead of one for each statement. This significantly improves speed on high latency connections when the statement cache is disabled or cold.
* `QueryFunc` method is replaced by `ForEachRow` functions.
* `BeginFunc` and `BeginTxFunc` methods are replaced by functions.
* `pgxpool.Pool` always uses lazy connect mode.
* Added tracing hooks for integration with tools like OpenTelemetry. This replaces the existing logging interface, but a compatibility adapter is provided in the `tracelog` package.

No more features or breaking changes are planned, but further testing may reveal necessary changes. So now's the time to test and request interface changes if necessary. My plan is for this to be the final alpha release.

**Comment by jackc** [comment]
pgx v5 alpha testing is now concluded. https://github.com/jackc/pgx/issues/1273 is the new issue for the beta.
